### PR TITLE
Shoaib/lp 145

### PR DESCRIPTION
**Pull request by shoaibsattar823**
@omairshamshir I have commited the changes for the admin usurping logics [LP-145].
Kindly review it.

### CODE REVIEW

File: lms/djangoapps/onboarding/views.py [hunk 1]
```diff
@@ -355,10 +356,41 @@ def update_account_settings(request):
...
+    org_id = user_extended_profile.organization_id
+    org_name = organization.label
+    prev_admin_email = organization.admin.email
+    send_admin_change_email(org_id, org_name, prev_admin_email, request.user.username, request.user.email)
+    return HttpResponseRedirect('/onboarding/account_settings/')
+
+
+@csrf_exempt
+def admin_change_confirmation(request, confirmation, username):
+    user = User.objects.get(username=username)
```
Comment (rosheen33) [inline lms/djangoapps/onboarding/views.py]: Kindly add DOC strings for views.

File: lms/djangoapps/onboarding/views.py [hunk 2]
```diff
@@ -355,10 +356,41 @@ def update_account_settings(request):
...
+        admin_email = organization.admin.email
+        if confirmation == "1":
+            organization.unclaimed_org_admin_email = None
+            organization.admin = user
+            organization.save()
+            send_admin_change_confirmation_email(org_name, admin_email, username, user.email, confirm=1)
+            return HttpResponseRedirect('/onboarding/account_settings/')
+        else:
+            send_admin_change_confirmation_email(org_name, admin_email, username, user.email, confirm=0)
+            return HttpResponseRedirect('/onboarding/account_settings/')
```
Comment (rosheen33) [inline lms/djangoapps/onboarding/views.py]: Can we replace above with this ?
```
        send_admin_change_confirmation_email(
            org_name, admin_email, username, user.email, confirm=1 if confirmation == '1' else 0
        )
        return HttpResponseRedirect('/onboarding/account_settings/')
```

File: lms/djangoapps/onboarding/urls.py [hunk 3]
```diff
@@ -19,4 +19,7 @@
     url(r"^organization_detail/$", views.org_detail_survey, name="org_detail_survey"), # signup step 4
     url(r"^delete_account/$", views.delete_my_account, name="delete_my_account"), # signup step 4
     url(r"^admin_activate/(?P<org_id>[^/]*)/(?P<activation_key>[^/]*)$", views.admin_activation, name="admin_activation"),
+    url(r"^admin_change_request/$", views.admin_change, name="admin_change"),
+    url(r"^admin_change_request/(?P<confirmation>[0-1])/(?P<username>[^/]*)$",
```
Comment (rosheen33) [inline lms/djangoapps/onboarding/urls.py]: Can we use the Hash Approach here as in **admin_activation** view ?

**Comment by rosheen33** [comment]
@muhammadnadeem Kindly review now.

File: lms/djangoapps/onboarding/views.py [hunk 4]
```diff
@@ -566,39 +507,68 @@ def admin_activation(request, org_id, activation_key):
 
     """
     context = {}
-    hash_key_obj = None
+    admin_activation = eval(request.GET.get('admin_activation', 'False'))
```
Comment (muhammadnadeem) [inline lms/djangoapps/onboarding/views.py]: can we use bool instead of eval?

File: lms/djangoapps/onboarding/views.py [hunk 5]
```diff
@@ -566,39 +507,68 @@ def admin_activation(request, org_id, activation_key):
...
 
     try:
-        hash_key_obj = OrganizationAdminHashKeys.objects.get(activation_hash=activation_key)
-        user_extended_profile = UserExtendedProfile.objects.get(user__email=hash_key_obj.suggested_admin_email)
+        hash_key = OrganizationAdminHashKeys.objects.get(activation_hash=activation_key)
+        admin_change_confirmation = True if eval(request.GET.get('confirm', u'False')) else False
```
Comment (muhammadnadeem) [inline lms/djangoapps/onboarding/views.py]: can we write it as
True if request.GET.get('confirm') else False?

File: lms/djangoapps/onboarding/views.py [hunk 6]
```diff
@@ -566,39 +507,68 @@ def admin_activation(request, org_id, activation_key):
...
+        user_profile = UserExtendedProfile.objects.get(user__email=hash_key.suggested_admin_email)
```
Comment (muhammadnadeem) [inline lms/djangoapps/onboarding/views.py]: it should be user_extended_profile as UserProfile is another model

File: lms/djangoapps/onboarding/views.py [hunk 7]
```diff
@@ -566,39 +507,68 @@ def admin_activation(request, org_id, activation_key):
...
+                if admin_change_confirmation:
+                    if user_profile.organization.admin == user:
+                        user_profile.organization.admin = None
+                        user_profile.organization.save()
+
+                    hash_key.organization.unclaimed_org_admin_email = None
+                    hash_key.organization.admin = user
+                    hash_key.organization.save()
+
+                    user_profile.organization = hash_key.organization
```
Comment (muhammadnadeem) [inline lms/djangoapps/onboarding/views.py]: remove duplicate code from 527 - 540
+ why are you assigning organization from has_key object?

File: lms/djangoapps/onboarding/forms.py [hunk 8]
```diff
@@ -675,6 +681,19 @@ def save(self, user=None, commit=True):
...
+                    send_admin_change_email(
+                        organization_to_assign.id,
+                        organization_to_assign.label,
+                        organization_to_assign.admin.email,
+                        hash_key, user.email, user.username
+                    )
+                except Exception as ex:
+                    log.info(ex.args)
+                    pass
+
```
[the remark this answers is not in the document]
Comment (rosheen33) [reply lms/djangoapps/onboarding/forms.py]: This section is been removed. now we are using ajax requests to send emails.

File: lms/djangoapps/onboarding/views.py [hunk 9]
```diff
@@ -507,13 +507,13 @@ def admin_activation(request, activation_key):
...
     context = {}
-    admin_activation = eval(request.GET.get('admin_activation', 'False'))
+    admin_activation = True if request.GET.get('admin_activation') == 'True' else False
 
     try:
         hash_key = OrganizationAdminHashKeys.objects.get(activation_hash=activation_key)
-        admin_change_confirmation = True if eval(request.GET.get('confirm', u'False')) else False
-        user_profile = UserExtendedProfile.objects.get(user__email=hash_key.suggested_admin_email)
-        user = user_profile.user
+        admin_change_confirmation = True if request.GET.get('confirm') == 'True' else False
```
Comment (muhammadnadeem) [inline lms/djangoapps/onboarding/views.py]: please execute the following expressions in python shell and see out out
bool('True')
bool('False')
bool(None)
bool('abc')

so if we simply write this as following it should work 
admin_change_confirmation = bool(request.GET.get('confirm'))

Comment (rosheen33) [reply lms/djangoapps/onboarding/views.py]: bool('True') = True
bool('False') = True
bool(None) = False
bool('abc') = True

Comment (rosheen33) [reply lms/djangoapps/onboarding/views.py]: bool('False') = True doesn't works correctly in our scenario.

Comment (muhammadnadeem) [reply lms/djangoapps/onboarding/views.py]: ok, please go for merge

**Pull request merged**